Created CreepStressUpdate base for creep fundamentals

tophmatthews opened the pull request:
Closes #11412

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

tophmatthews committed:
Created CreepStressUpdate base for creep fundamentals
Closes #11412

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/185113/) on 7288d39 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11418/site)

This comment will be updated on new commits.

bwspenc commented:
I'd like to hear your take on this, but once I saw how little code the individual creep models need to supply to compute the tangent operator, I'm not convinced that this is actually that helpful. Adding this extra level of inheritance doesn't really save duplication of that much code, and it does add complexity.

tophmatthews commented:
Yeah, but I like how the actual implementation of any creep model focuses only on the model itself, especially when dealing with the models in BISON. These details, not just the tangent, are just code duplication.

tophmatthews commented:
Ideally, you only have computeResidual and computeDerivative in any inherited classes

tophmatthews commented on modules/tensor_mechanics/src/materials/CreepStressUpdateBase.C:
```diff
@@ -0,0 +1,66 @@
+void
+CreepStressUpdateBase::computeStressFinalize(const RankTwoTensor & plasticStrainIncrement)
+{
+  _creep_strain[_qp] += plasticStrainIncrement;
```
@bwspenc This seems error prone to me. Shouldn't this just be `_creep_strain[_qp] += _creep_strain_old[_qp] + plasticStrainIncrement`?

bwspenc replied:
You are correct.  If you look at the BISON Zry creep models, they do it the way you say. Also, change the name `plasticStrainIncrement` -> `inelastic_strain_increment`.

bwspenc commented:
OK, I see that. It looks like just by switching those creep models over to derive from this class, they will automatically get their nonlinear tangent operator computed. Is that right?  If so, that's a really good thing!

tophmatthews commented on modules/tensor_mechanics/src/materials/PowerLawCreepStressUpdate.C:
```diff
@@ -79,8 +49,9 @@ PowerLawCreepStressUpdate::computeStressInitialize(const Real /*effective_trial_
+  mooseAssert(_t >= _start_time, "Time must be greater than start time");
   _exp_time = std::pow(_t - _start_time, _m_exponent);
 
   _creep_strain[_qp] = _creep_strain_old[_qp];
```
@bwspenc Is this step necessary? If it is generic, it should be moved to the base class.

bwspenc replied:
I don't think it is necessary.

tophmatthews replied:
Ah, I see. Testing it out, it's necessary if you do `_creep_strain[_qp] += plasticStrainIncrement;`. I'll take out both of these.

dschwen replied:
If you don't set the current material property it will retain the value from the previous element. I just ran into the same thing in some code I'm working on. Just be careful about this because there will not be a runtime error and you will only notice this when looking at your results very carefully (you'll see an odd horizontal streaking in your properties).

tophmatthews replied:
Huh. Do you mean it'll retain the value from the previous qp altogether? What is the initial value of a material property before being calculated? Where does `propagateQpStatefulPropertiesRadialReturn` fit in?

dschwen replied:
Ah, I think my case was slightly different as I was setting the value of a non-stateful property using the data from a stateful property (hence the streaking). Here teh error would be even more subtle as the current material data would be cycled around and would probably hold the previous _older_ data. In any case it its necessary to write to the current data in _all_ cases. I haven't followed the code flow here so take it with a grain of salt, but you cannot do any of teh following

* not write to a property data entry (stateful or not does not matter)
* read before you write (e.g. `+=` operator)

bwspenc replied:
@tophmatthews If you do what you proposed, it will be handled correctly in all cases. Look at the models in BISON -- they do it the way you proposed.

tophmatthews replied:
Yeah, that's what I thought. I don't think creep_strain is used before it is set.

tophmatthews commented on modules/tensor_mechanics/src/materials/CreepStressUpdateBase.C:
```diff
@@ -0,0 +1,66 @@
+Real
+CreepStressUpdateBase::computeStressDerivative(const Real effective_trial_stress, const Real scalar)
+{
+  return -(computeDerivative(effective_trial_stress, scalar) + 1.0) / _three_shear_modulus;
```
@bwspenc This is really the only calculated piece that the creep models need to calcualte the tangent. This is the derivative of the creep strain wrt to the von mises stress, which (I think) all creep models use the same.

tophmatthews commented on modules/tensor_mechanics/include/materials/CreepStressUpdateBase.h:
```diff
@@ -0,0 +1,68 @@
+   */
+  virtual TangentCalculationMethod getTangentCalculationMethod() override
+  {
+    return TangentCalculationMethod::PARTIAL;
```
@bwspenc This is the other piece that is needed to "turn on" the tangent calculation. All creep models will be `PARTIAL` by definition.

bwspenc commented on modules/tensor_mechanics/src/materials/CreepStressUpdateBase.C:
```diff
@@ -0,0 +1,66 @@
+  return params;
+}
+
+CreepStressUpdateBase::CreepStressUpdateBase(const InputParameters & parameters)
```
Do you think it makes sense to name this `J2CreepStressUpdateBase` or something, just to make it clear that it is only applicable to J2 models?

tophmatthews replied:
Isn't that already decided by RadialReturnStressUpdate?

bwspenc replied:
Sure. I'm just thinking that someone with a non-J2 model might pick this up and think they can derive their model from it. They ought to quickly see that they can't, but the name you have implies that it's more general than it is.

bwspenc replied:
Maybe `RadialReturnCreepStressUpdateBase`? Not a super pretty name, but it doesn't matter too much since it's not exposed in the input files.

tophmatthews replied:
Yeah, I like `RadialReturnCreepStressUpdateBase` better, it makes is clear. A mouthful, but very clear.

bwspenc replied:
We had a very heated discussion about the `Update` in the name. The idea was to make it clear that these models couldn't compute the stress by themselves. 

What I'm still lobbying for is getting rid of `Compute` in front of all the models!

tophmatthews replied:
Ah! I thought it was to separate it with solid mechanics.

tophmatthews commented:
@bwspenc alright, renamed it and fixed it all up what we talked about. The only other weirdness is the `start_time` parameter, which I think is supposed to match the executioner block if a `start_time` is used. I threw an assert on it to prevent `std::pow` failures.

I also removed `_max_creep_incr` which I think is left over from old ways of limiting the strain rate.

bwspenc commented on modules/tensor_mechanics/include/materials/PowerLawCreepStressUpdate.h:
```diff
@@ -20,59 +19,51 @@ InputParameters validParams<PowerLawCreepStressUpdate>();
 
 /**
  * This class uses the Discrete material in a radial return isotropic creep
```
Could you remove all references to a Discrete material? At one point, we had a special Material base class called DiscreteMaterial that accomplished the same thing as what we now do by setting `update=false`. You could call it an Update material or a compute=false material or something.

bwspenc reviewed (changes requested):
Really close -- just a couple of little things as noted.

tophmatthews commented on modules/tensor_mechanics/src/materials/PowerLawCreepStressUpdate.C:
```diff
@@ -79,11 +49,10 @@ PowerLawCreepStressUpdate::computeStressInitialize(const Real /*effective_trial_
-    _exponential = 1;
+    _exponential = 1.0;
 
+  mooseAssert(_t >= _start_time, "Time must be greater than start time");
```
It's done once per timestep, but yeah I can change to error. I wonder if there is a way to check what the executioner param is set to?

tophmatthews replied:
Hmmm...there is `Transient::getStartTime`. Maybe it'd be better just to pull from that instead of allowing the user to set this?

tophmatthews replied:
I can access the executioner start_time via `_app.getStartTime()`. I was thinking of just setting `_start_time = _app.getStartTime()` and then deprecating the start_time param. I don't think it's even tested now anywhere. If you think it best to keep the option to have a different start_time, I can do that, though I don't much see a case for it.

bwspenc replied:
Sure, do that.

tophmatthews commented:
@bwspenc Alright, let's try this one. I removed discrete references in favor of "stress update". I set `_start_time(_app.getStartTime())`, and deprecated the input param asking for start_time. I also added a test that changing the start time gives the same answer.

tophmatthews commented:
@bwspenc Most bison models declare a creep_rate material property. We could easily implement that here too.

bwspenc commented on modules/tensor_mechanics/src/materials/PowerLawCreepStressUpdate.C:
```diff
@@ -8,68 +8,42 @@
-
-void
-PowerLawCreepStressUpdate::initQpStatefulProperties()
+    _start_time(_app.getStartTime())
```
The way you have it, `start_time` will be ignored if set by the user. This would use it if it was set, otherwise use the app's start time:
```
_start_time(parameters.isParamSetByUser("start_time") ? getParam<Real>("start_time") : _app.getStartTime())
```
The problem is that the default would no longer be zero, as is claimed, and as it used to be.

tophmatthews replied:
Oh I see. This will retain functionality I suppose. Removing the default, I can then use `isParamValid()`. would that be better?

bwspenc replied:
I don't see a super clean way to avoid changing old behavior. Maybe do what I suggested above, and then just below in the body of the constructor, do this:
```
if (_m_exponent != 0.0 && _app.getStartTime() != 0.0 && !isParamSetByUser("start_time"))
  mooseError("The default start_time in PowerLawCreepStressUpdate is the executioner's start time, rather than 0");
```
Never mind -- that doesn't work either, because then it will never run if your executioner starts at a nonzero time.  Maybe just leave this whole start_time thing alone since it's unrelated to what you really want to do here?

tophmatthews replied:
Checkout what I added. I think that gets the root of the reason it is there. If the user does tries to add their own start_time, it'll get a deprecated warning. Otherwise, it'll be set by the executioner's start time. I suppose I could throw a check to ensure that the user start time is greater than the executioner's start time. That'll take care of all cases.

Really, I want to just yank the input param as is...

bwspenc commented on modules/tensor_mechanics/src/materials/PowerLawCreepStressUpdate.C:
```diff
@@ -8,68 +8,41 @@
-
-void
-PowerLawCreepStressUpdate::initQpStatefulProperties()
+    _start_time(isParamValid("start_time") ? getParam<Real>("start_time") : _app.getStartTime())
```
This works, except that if someone had a model that started at a nonzero time and did not specify start_time, it would run differently now.

tophmatthews replied:
Yeah that's true...

bwspenc commented on modules/tensor_mechanics/src/materials/RadialReturnCreepStressUpdateBase.C:
```diff
@@ -0,0 +1,69 @@
+{
+  InputParameters params = validParams<RadialReturnStressUpdate>();
+
+  params.addCoupledVar("temperature", 0.0, "Coupled temperature");
```
Also in the stupid defaults department, temperature shouldn't have defaulted to 0...

bwspenc replied:
That's easily fixed by making it required. Just change this to `addRequiredCoupledVar`. If they weren't specifying it before, they'll get an error, and if they really wanted it to be zero, they can say `temperature=0` in the input file.

tophmatthews replied:
Yeah okay, that'll change the _has_temp behavior which is protecting a divide by 0

tophmatthews replied:
I think that can get yanked as well, and do a zero check on temperature instead. It's safer that way anyways in case the model returns a coupled zero temperature value.

tophmatthews replied:
whoa wait a second, it's not even calculating the temperature exponential on the fly anyways!! It's only doing it at the beginning of the timestep!

tophmatthews replied:
Defaulting temperature to zero seems okay too. It's just saying that creep strain is temperature independent?

tophmatthews replied:
Most of the tests now assume temperature is 0.

tophmatthews replied:
The rather, maybe a default of 0 is silly, but making it required doesn't seem the right way either.

tophmatthews commented:
@bwspenc Added an error if `_start_time < _app.getStartTime()`, and removed temperature defaults. I didn't make it required though. It is protected in PowerLawCreepStressUpdate by checking `_has_temp`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/185849/) on 0929140 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11418/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11418/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 116c246b12aaa8d80920103946683971413cd162`

bwspenc commented on modules/tensor_mechanics/src/materials/RadialReturnCreepStressUpdateBase.C:
```diff
@@ -0,0 +1,69 @@
+  : RadialReturnStressUpdate(parameters),
+    _creep_prepend(getParam<std::string>("creep_prepend")),
+    _has_temp(isCoupled("temperature")),
+    _temperature(_has_temp ? coupledValue("temperature") : _zero),
```
It may have made sense for that particular creep model to default temperature to zero, but that's a bad idea in general for all creep models that will potentially derive from this class. Since temperature isn't even used at this level, I'd say move it up to the power law creep model and leave its default behavior as it was.

tophmatthews replied:
Sure, it is a model parameter anyways. Keep the default at zero in the power law model?

bwspenc replied:
I would prefer to make it required, but I'm OK either way. I guess to minimize the impact of this commit on code behavior, it might be best to leave it alone.

Making it required isn't terribly dangerous to do because the worst thing that will happen is that someone's input file won't work, and then they'll go and add the missing parameter.  The nice thing is that you can supply a number (rather than a variable) for a parameter that accepts a coupled variable, so you can just set 'temperature=0' in the input file.

tophmatthews replied:
Yeah, that behavior isn't even valid now since there is a `isCoupled` guard for temperature.

tophmatthews commented:
okay...trying again! Moved temperature back to `PowerLawCreepStressUpdate`. Changed the `_has_temp` guard to `parameters.isParamSetByUser` instead of `isCoupled` to allow numbers to be used in place of variables in the input file.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/185893/) on 17b2733 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11418/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11418/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 116c246b12aaa8d80920103946683971413cd162`

tophmatthews commented:
Alright, I have now it'll throw a warning to prevent taking a non-integer power of a negative number.

We're probably over thinking this way too much...I bet no one uses it except to match the executioner start time...

bwspenc commented on modules/tensor_mechanics/src/materials/PowerLawCreepStressUpdate.C:
```diff
@@ -8,68 +8,46 @@
+  if (_start_time < _app.getStartTime() && (std::trunc(_m_exponent) != _m_exponent))
+    paramError("start_time",
+               "Start time must be equal to or greater than the Executioner start_time if a "
+               "non-integer m_exponent is used");
```
Shouldn't it be nonzero, rather than non-integer?

tophmatthews replied:
That would be the easy way, but I thought want to maintain previous abilities. You can still take an integer power of a negative number.

bwspenc replied:
Oh... that's really over-thinking it!

tophmatthews replied:
Say the word, I'll yank it all out and no one (probably) would notice!

bwspenc replied:
No leave it alone. I think we've beat this thing to death!
